timdex_dataset_api/embeddings.py

```diff
@@ -379,21 +405,66 @@ def _build_query(
         """Build SQL query using SQLAlchemy.
 
         The method returns a SQL query string, which SQLAlchemy executes to
-        fetch results.
+        fetch results. Always joins to metadata.records to enable filtering
+        by metadata columns (source, run_date, run_type, action, run_timestamp).
         """
-        sa_table = self.get_sa_table(table)
+        embeddings_table = self.get_sa_table(table)
+        metadata_table = self.timdex_dataset.metadata.get_sa_table("records")
+
+        # create SQL statement with join to metadata.records
+        join_condition = and_(
+            embeddings_table.c.timdex_record_id == metadata_table.c.timdex_record_id,
+            embeddings_table.c.run_id == metadata_table.c.run_id,
+            embeddings_table.c.run_record_offset == metadata_table.c.run_record_offset,
+        )
 
-        # create SQL statement object
-        # select named columns
+        # select specific columns or default to all from embeddings + metadata
         if columns:
-            stmt = select(*sa_table.c[*columns]).select_from(sa_table)
+            embeddings_cols = []
+            metadata_cols = []
+
+            for col_name in columns:
+                if col_name in TIMDEX_DATASET_EMBEDDINGS_SCHEMA.names:
+                    embeddings_cols.append(embeddings_table.c[col_name])
+                elif col_name in METADATA_SELECT_FILTER_COLUMNS:
+                    metadata_cols.append(metadata_table.c[col_name])
+                else:
+                    raise ValueError(f"Invalid column: {col_name}")
```

Comment:
Because columns for filtering and selecting can be found on the record metadata schema _or_ the embedding schema, we get a list of those columns here as SQLAlchemy objects.  Then, we can use these lists in filtering and selection below.